File: fixtures/scale/scala/large/src/cryptobench/ciphers/variantbeaufort/VariantBeaufortCipher.scala
package cryptobench.ciphers.variantbeaufort

import cryptobench.core.Cipher
import cryptobench.util.Alphabet

/** Subtracts the keyword letter on encrypt, adds it back on decrypt. */
final class VariantBeaufortCipher(key: VariantBeaufortKey) extends Cipher {

  override def name: String = "variantbeaufort"

  override def encrypt(plaintext: String): String = transform(Alphabet.clean(plaintext), forward = true)

  override def decrypt(ciphertext: String): String = transform(Alphabet.clean(ciphertext), forward = false)

  private def transform(text: String, forward: Boolean): String = {
    val sb = new StringBuilder
    for (i <- text.indices) {
      val x = Alphabet.indexOf(text.charAt(i))
      val k = Alphabet.indexOf(key.keyCharAt(i))
      sb.append(Alphabet.charAt(if (forward) x - k else x + k))
    }
    sb.toString
  }
}

File: fixtures/scale/dart/large/lib/ciphers/variantbeaufort/variantbeaufort_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import 'variantbeaufort_key.dart';

/// Subtracts the keyword letter on encrypt, adds it back on decrypt.
class VariantBeaufortCipher implements Cipher {
  final VariantBeaufortKey key;

  VariantBeaufortCipher(this.key);

  @override
  String name() => 'variantbeaufort';

  @override
  String encrypt(String plaintext) => _transform(cleanAlphabet(plaintext), true);

  @override
  String decrypt(String ciphertext) => _transform(cleanAlphabet(ciphertext), false);

  String _transform(String text, bool forward) {
    final sb = StringBuffer();
    for (var i = 0; i < text.length; i++) {
      final x = indexOf(text[i]);
      final k = indexOf(key.keyCharAt(i));
      sb.write(charAt(forward ? x - k : x + k));
    }
    return sb.toString();
  }
}
